### PR TITLE
Module sub targets

### PR DESCRIPTION
**Pull request by aherrmann**
Closes https://github.com/MercuryTechnologies/the-culture-repo/issues/143
cc @wavewave 

Adds per-module sub-targets for .o and .hi outputs.

**query the sub-targets**
```
$ buck2 audit subtargets //backend/src:backend_sql
...
root//backend/src:backend_sql[static][interfaces][Sql.Gucci] (prelude//platforms:default#904931f735703749)
...
root//backend/src:backend_sql[static][objects][Sql.Gucci] (prelude//platforms:default#904931f735703749)
...
root//backend/src:backend_sql[static-pic][interfaces][Sql.Gucci] (prelude//platforms:default#904931f735703749)
...
root//backend/src:backend_sql[static-pic][objects][Sql.Gucci] (prelude//platforms:default#904931f735703749)
...
root//backend/src:backend_sql[shared][interfaces][Sql.Gucci] (prelude//platforms:default#904931f735703749)
...
root//backend/src:backend_sql[shared][objects][Sql.Gucci] (prelude//platforms:default#904931f735703749)
...
```

**build the sub-targets**
```
$ buck2 build //backend/src:backend_sql[static][interfaces][Sql.Gucci] --show-output
...
root//backend/src:backend_sql[static][interfaces][Sql.Gucci] buck-out/v2/gen/root/904931f735703749/backend/src/__backend_sql__/mod-static/Sql/Gucci.hi

$ buck2 build //backend/src:backend_sql[shared][objects][Sql.Gucci] --show-output
...
root//backend/src:backend_sql[shared][objects][Sql.Gucci] buck-out/v2/gen/root/904931f735703749/backend/src/__backend_sql__/mod-shared/Sql/Gucci.dyn_o
```

- **Add per-module sub-targets for .o and .hi**
- **Capture matchink link-style and profiling-mode**

### CODE REVIEW

File: haskell/haskell.bzl [hunk 1]
```diff
@@ -1205,12 +1211,37 @@ def haskell_binary_impl(ctx: AnalysisContext) -> list[Provider]:
...
+        compiled = compiled,
+        link_style = link_style,
+        enable_profiling = enable_profiling,
+    ))
+
     providers = [
-        DefaultInfo(default_output = output),
+        DefaultInfo(
+            default_output = output,
+            sub_targets = sub_targets,
```
Comment (aherrmann) [inline haskell/haskell.bzl:1224]: For some reason this does not seem to work on `haskell_binary` targets, yet. The following lists no sub-targets:
```
$ buck2 audit subtargets //backend/one-off-task:one-off-task
```
Per print debugging the Starlark code this should list per-module objects and interfaces sub-targets.

**Comment by wavewave** [comment]
Oh nice!

File: haskell/haskell.bzl (hunk 1 above)
Comment (aherrmann) [reply haskell/haskell.bzl:1224]: Nevermind, I just missed the `-hs` suffix in the target name.
```
$ buck2 audit subtargets //backend/one-off-task:one-off-task-hs
```
Works as expected.

**Comment by aherrmann** [comment]
This combined with the cross-package module dependencies means that we can avoid linking actions altogether when building objects. E.g.
```
$ buck2 build //backend/one-off-task:one-off-task-hs[objects][TaskMain]
```
does not involve any linking of any package libraries.

**Review by avdv — APPROVED** [review]
Looks good! Thank you.

**Pull request merged**